Elide bignum operations

paulapatience opened the pull request.

yitzchak commented on code/math/implementation.lisp:
```diff
@@ -1,17 +1,173 @@
+     (64 (if (eql count 1)
+             `(unsigned-byte 64)
+             `(simple-array (unsigned-byte 64) (,count))))
+     (128 `(simple-array (unsigned-byte 64) (,(ash count 1)))))
```
This should probably be ``(128 `(unsigned-byte ,(* arithmetic-size count))))``. We are currently only need arithmetic size 128 on ECL, but having the 128-bit routines not have a no-bignum version will greatly simplify things, since we would need to have a custom LDB, ASH, etc.

yitzchak commented on code/math/implementation.lisp:
```diff
@@ -38,15 +189,101 @@
          (ash p ,(- size)))))
 
 (defun round-to-odd/32 (g cp)
```
Let's put these in the same order as floor-multiply (swap g and cp) and use the same naming scheme.

yitzchak commented on code/math/implementation.lisp:
```diff
@@ -38,15 +189,101 @@
+        (ash p -32)))
+  #-quaviver/bignum-elision
+  (progn
+    #-(or ecl cmucl) (%round-to-odd-1 g cp 32)
```
I'd ditch the `progn` and just do 
```
 #-(and (not quaviver/bignum-elision) (or ecl cmucl)) (%round-to-odd-1 g cp 32)
 #+(and (not quaviver/bignum-elision) (or ecl cmucl)) (%round-to-odd-2 g cp 32)
```

Yes, its a little more verbose, but it puts all of the conditional expressions on the same level and is bit clearer.

yitzchak commented on code/schubfach/implementation.lisp:
```diff
@@ -22,7 +22,7 @@
                       (h (+ exponent 1 (quaviver/math:floor-log-expt 2 10 (- k))))
                       (expt10 (,expt10 k)))
-                 (declare (type (unsigned-byte ,(ash arithmetic-size 1))
+                 (declare (type (quaviver/math::word ,arithmetic-size 2)
```
`word` needs to be exported. Maybe name it arithmetic-word instead?

paulapatience committed:
math: constrain declared FTYPEs to return one value

A declared result type of T is effectively (VALUES T &REST T).
To constrain it to a single value, the syntax is (VALUES T &OPTIONAL).

* code/math/implementation.lisp (round-to-odd/32, round-to-odd/64)
(round-to-odd/128, expt10/32, expt10/64, expt10/128, floor-log-expt)
(ceiling-log-expt): Declare result types with the (VALUES T &OPTIONAL)
syntax.

paulapatience committed:
liebler: add some type declarations

* code/liebler/implementation.lisp (%liebler): Add some type
declarations.

Co-authored-by: Tarn W. Burton <twburton@gmail.com>

paulapatience committed:
math: introduce ARITHMETIC-WORD type

It uses the anticipatory feature QUAVIVER/MATH/SMALLNUM.

* code/math/implementation.lisp (arithmetic-word): New type.
(round-to-odd/32, round-to-odd/64, round-to-odd/128, expt10/32)
(expt10/64, expt10/128): Use it in FTYPE declarations.
* code/math/packages.lisp (#:quaviver/math): Export ARITHMETIC-WORD.

Co-authored-by: Tarn W. Burton <twburton@gmail.com>

paulapatience committed:
liebler, schubfach, dragonbox: use ARITHMETIC-WORD

* code/liebler/implementation.lisp (%liebler):
* code/schubfach/implementation.lisp (%schubfach): Use ARITHMETIC-WORD,
declaring any EXPT10 variables DYNAMIC-EXTENT.
* code/dragonbox/implementation.lisp (floor-by-expt10-divisible-p)
(floor-by-expt10-small, floor-by-expt10, compute-mul)
(compute-mul-parity, %nearest, %directed): Same, and replace previous
uses of EXPT10-SIZE with (ASH ARITHMETIC-SIZE 1).

paulapatience committed:
math, liebler, schubfach: swap ROUND-TO-ODD parameter order

In anticipation for upcoming math routines, where the principal
arithmetic-sized parameter is first.

* code/math/implementation.lisp (%round-to-odd-1, %round-to-odd-2)
(round-to-odd/32, round-to-odd/64, round-to-odd/128): Swap parameter
order.
* code/liebler/implementation.lisp (%liebler):
* code/schubfach/implementation.lisp (%schubfach): Update.

paulapatience committed:
math, liebler, schubfach: rename ROUND-TO-ODD

So that the sizes of all the parameters are apparent.

* code/math/implementation.lisp (round-to-odd/32, round-to-odd/64)
(round-to-odd/128): Rename to...
(round-to-odd/32-64, round-to-odd/64-128, round-to-odd/128-256): ...
these.
* code/math/packages.lisp (#:quaviver/math):
* code/liebler/implementation.lisp (quaviver:integer-float):
* code/schubfach/implementation.lisp (quaviver:float-integer): Update.

paulapatience committed:
math, dragonbox: extract FLOOR-MULTIPLY operations

From quaviver/dragonbox's COMPUTE-MUL operations.

* code/math/implementation.lisp: Add copyright header.
(%floor-multiply/n-2nq2n, %floor-multiply/evenp/n-2nq2n): New macros,
extracted from code/dragonbox/implementation.lisp.
(floor-multiply/32-64q64, floor-multiply/64-128q128)
(floor-multiply/evenp/32-64q64, floor-multiply/evenp/64-128q128): New
functions.
* code/math/utility.lisp: Add copyright header.
* code/math/packages.lisp: Same.
(#:quaviver/math): Export new functions.
* code/dragonbox/implementation.lisp (compute-mul, compute-mul-parity):
Remove macros now found in quaviver/math.
(beta-bounds/nearest/shorter-interval)
(beta-bounds/nearest/normal-interval, beta-bounds/left-closed-directed)
(beta-bounds/right-closed-directed): New functions.
(%nearest, %directed, quaviver:float-integer): Update to use
arithmetic-size–specific FLOOR-MULTIPLY and FLOOR-MULTIPLY/EVENP
functions.
* quaviver.asd (quaviver/math): Update license.

paulapatience committed:
math, dragonbox: introduce HI operations

For extracting the high bits of a word. These operations allow some
simplifications in quaviver/dragonbox's nearest shorter interval case
and in the computations of DELTAI.

* code/math/implementation.lisp (hi/64, hi/hi64/128): New functions.
* code/math/packages.lisp (#:quaviver/math): Export them.
* code/dragonbox/implementation.lisp (%nearest, %directed)
(quaviver:float-integer): Use them.

paulapatience committed:
math, dragonbox: introduce smallnum optimizations

Mostly ported from the reference implementation of Dragonbox.

They are activated by default only on SBCL, as on other implementations
they tend to be slower than bignum operations.

* code/math/implementation.lisp (+/64-64, +/128-64, */32-64/hi64)
(*/32-64/lo64, */64-64, */64-64/hi64, */64-64/lo64, */64-128/hi128)
(*/64-128/lo128): New functions, mostly ported from reference Dragonbox.
(hi/hi64/128): New smallnum optimization and compiler macro.
(round-to-odd/32-64, round-to-odd/64-128): New smallnum optimizations.
(floor-multiply/32-64q64, floor-multiply/64-128q128)
(floor-multiply/evenp/32-64q64, floor-multiply/evenp/64-128q128): New
smallnum optimizations from reference Dragonbox.
(*expt10/values/64*): Add smallnum version.
* code/math/packages.lisp: Push QUAVIVER/MATH/SMALLNUM feature on SBCL.
* code/dragonbox/implementation.lisp (floor-by-expt10): New smallnum
optimization from reference Dragonbox.

paulapatience committed:
schubfach: remove spurious character 'l'

* code/schubfach/implementation.lisp (%schubfach): Remove spurious
character.
